Adopt even more smart pointers in DOM code

cdumez opened the pull request:
#### 0e12ca3ef084d8fe6459fb909e7a784d879adec6
<pre>
Adopt even more smart pointers in DOM code
<a href="https://bugs.webkit.org/show_bug.cgi?id=263505">https://bugs.webkit.org/show_bug.cgi?id=263505</a>

Reviewed by Darin Adler.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWindowProxy const):
(WebCore::Document::checkedSelection):
(WebCore::Document::checkedSelection const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentParser.cpp:
(WebCore::DocumentParser::protectedDocument const):
* Source/WebCore/dom/DocumentParser.h:
* Source/WebCore/dom/PendingScript.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::Range):
(WebCore::Range::~Range):
(WebCore::Range::protectedOwnerDocument):
(WebCore::Range::updateAssociatedSelection):
(WebCore::Range::updateAssociatedHighlight):
(WebCore::Range::updateDocument):
(WebCore::childOfCommonRootBeforeOffset):
(WebCore::Range::protectedStartContainer const):
(WebCore::Range::protectedEndContainer const):
(WebCore::Range::processContents):
(WebCore::processContentsBetweenOffsets):
(WebCore::processAncestorsAndTheirSiblings):
(WebCore::Range::insertNode):
(WebCore::Range::toString const):
(WebCore::Range::createContextualFragment):
(WebCore::Range::checkNodeOffsetPair):
(WebCore::Range::cloneRange const):
(WebCore::Range::setStartAfter):
(WebCore::Range::setEndBefore):
(WebCore::Range::setEndAfter):
(WebCore::Range::selectNode):
(WebCore::Range::surroundContents):
(WebCore::Range::setStartBefore):
(WebCore::Range::updateRangeForParentlessNodeMovedToNewDocument):
(WebCore::boundaryTextNodesMerged):
(WebCore::boundaryTextNodesSplit):
(WebCore::Range::expand):
(WebCore::Range::getClientRects const):
(WebCore::Range::getBoundingClientRect const):
(WebCore::setBothEndpoints):
(WebCore::makeSimpleRange):
(WebCore::createLiveRange):
* Source/WebCore/dom/Range.h:
* Source/WebCore/dom/RangeBoundaryPoint.h:
(WebCore::RangeBoundaryPoint::set):
(WebCore::RangeBoundaryPoint::setToAfterNode):
* Source/WebCore/dom/RawDataDocumentParser.h:
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/ScopedEventQueue.cpp:
(WebCore::ScopedEventQueue::dispatchEvent const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ref const):
(WebCore::ScriptElement::deref const):
(WebCore::ScriptElement::ref): Deleted.
(WebCore::ScriptElement::deref): Deleted.
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/ScriptRunner.cpp:
(WebCore::ScriptRunner::~ScriptRunner):
(WebCore::ScriptRunner::queueScriptForExecution):
(WebCore::ScriptRunner::resume):
(WebCore::ScriptRunner::notifyFinished):
(WebCore::ScriptRunner::timerFired):
* Source/WebCore/dom/ScriptRunner.h:
* Source/WebCore/dom/ScriptableDocumentParser.cpp:
(WebCore::ScriptableDocumentParser::scriptsWaitingForStylesheetsExecutionTimerFired):
* Source/WebCore/dom/ScriptedAnimationController.cpp:
(WebCore::ScriptedAnimationController::registerCallback):
(WebCore::ScriptedAnimationController::cancelCallback):
(WebCore::ScriptedAnimationController::serviceRequestAnimationFrameCallbacks):
(WebCore::ScriptedAnimationController::scheduleAnimation):
(WebCore::ScriptedAnimationController::protectedDocument):
* Source/WebCore/dom/ScriptedAnimationController.h:
* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::inheritPolicyContainerFrom):
(WebCore::SecurityContext::checkedContentSecurityPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::insertedIntoAncestor):
(WebCore::ShadowRoot::removedFromAncestor):
(WebCore::ShadowRoot::moveShadowRootToNewParentScope):
(WebCore::ShadowRoot::removeAllEventListeners):
(WebCore::ShadowRoot::findAssignedSlot):
(WebCore::ShadowRoot::assignedNodesForSlot):
(WebCore::assignedShadowRootsIfSlotted):
(): Deleted.
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::makeRangeSelectingNode):
* Source/WebCore/dom/SimulatedClick.cpp:
(WebCore::simulateMouseEvent):
(WebCore::simulateClick):
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::addSlotElementByName):
(WebCore::NamedSlotAssignment::didChangeSlot):
(WebCore::NamedSlotAssignment::assignSlots):
(WebCore::ManualSlotAssignment::addSlotElementByName):
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange):
(WebCore::ManualSlotAssignment::didRemoveManuallyAssignedNode):
* Source/WebCore/dom/StaticRange.cpp:
(WebCore::StaticRange::create):
* Source/WebCore/dom/StringCallback.cpp:
(WebCore::StringCallback::scheduleCallback):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::setInlineStyleFromString):
(WebCore::StyledElement::invalidateStyleAttribute):
(WebCore::StyledElement::addSubresourceAttributeURLs const):
* Source/WebCore/dom/TagCollection.cpp:
(WebCore::TagCollectionNS::~TagCollectionNS):
(WebCore::TagCollection::~TagCollection):
(WebCore::HTMLTagCollection::~HTMLTagCollection):
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::splitText):
(WebCore::Text::replaceWholeText):
(WebCore::Text::virtualCreate):
(WebCore::Text::updateRendererAfterContentChange):
(WebCore::Text::setDataAndUpdate):
* Source/WebCore/dom/TextDecoderStreamDecoder.cpp:
(WebCore::TextDecoderStreamDecoder::decode):
(WebCore::TextDecoderStreamDecoder::flush):
(WebCore::TextDecoderStreamDecoder::protectedTextDecoder):
* Source/WebCore/dom/TextDecoderStreamDecoder.h:
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.h:

Canonical link: <a href="https://commits.webkit.org/269644@main">https://commits.webkit.org/269644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3243c70fe144d5b475fd336d741516636a53307

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23147 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1232 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25065 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23413 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2678 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23685 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22256 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/804 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20070 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25914 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20956 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27120 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21075 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21220 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24988 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/649 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18421 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/580 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5521 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1058 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/856 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Range.cpp:
```diff
@@ -831,7 +844,7 @@ ExceptionOr<void> Range::selectNodeContents(Node& node)
 ExceptionOr<void> Range::surroundContents(Node& newParent)
 {
-    Ref<Node> protectedNewParent(newParent);
+    Ref protectedNewParent(newParent);
```
Usually I prefer the `=` style.

darinadler commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -125,8 +125,10 @@ Node::InsertedIntoAncestorResult ShadowRoot::insertedIntoAncestor(InsertionType
     DocumentFragment::removedFromAncestor(removalType, oldParentOfRemovedTree);
-    if (removalType.disconnectedFromDocument)
+    if (removalType.disconnectedFromDocument) {
+        // Unable to ref the document as it may have started destruction.
```
How does the static analyzer deal with this? It keeps coming up so I am wondering what our systematic way of dealing with the safety of such code should be.

cdumez replied:
The static analyzer had way too many issues to be truly helpful the last time I used it. My understanding is that Ryosuke is trying to improve it.

Ryosuke's plan for these particular cases was to allow ref'ing objects that started running their destructor:
- https://github.com/WebKit/WebKit/pull/8748

However, the patch was rejected and I do know that we have an alternative plan yet. I plan on eventually looking at refactoring the code to see if some of this code can be stopped from running when the document has started destruction, which is why I am annotating them with a comment.

I also heard that we may get support for pragmas to have the analyzer ignore part of the code in the future (but likely not soon).

darinadler commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -522,8 +524,9 @@ void ManualSlotAssignment::slotManualAssignmentDidChange(HTMLSlotElement& slot,
 
-    RenderTreeUpdater::tearDownRenderersAfterSlotChange(*shadowRoot.host());
-    shadowRoot.host()->invalidateStyleForSubtree();
+    RefPtr shadowRootHost = shadowRoot.host();
```
Since this is immediately unconditionally dereferenced I think we should use `Ref` instead of `RefPtr`. What do you think?

darinadler commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -543,10 +546,11 @@ void ManualSlotAssignment::slotManualAssignmentDidChange(HTMLSlotElement& slot,
 {
     ASSERT(slot.containingShadowRoot() == &shadowRoot);
-    ASSERT_UNUSED(node, node.parentNode() == shadowRoot.host());
+    RefPtr shadowRootHost = shadowRoot.host();
```
Since this is immediately unconditionally dereferenced I think we should use `Ref` instead of `RefPtr`. What do you think?

darinadler commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -111,7 +112,7 @@ class DragCaretController : private CaretBase {
 };
 
-class FrameSelection final : private CaretBase, public CaretAnimationClient {
+class FrameSelection final : private CaretBase, public CaretAnimationClient, public CanMakeCheckedPtr {
```
Originally we thought of these sub-objects of `Frame` as not having their own lifetime. Thus we’d really just want to forward the checking to `Frame` instead of having separate checking of the sub-objects. Same for strong references and weak references. I think an idiom like that would be more efficient and just as safe.

cdumez replied:
FrameSelection is a bit complicated. It is normally owned by a Document but not necessarily. We have the following code in accessibility:
```
    FrameSelection selection;
    selection.setSelection(position);
    selection.modify(FrameSelection::Alteration::Extend, SelectionDirection::Right, TextGranularity::LineBoundary);
    return selection.selection();
```

darinadler replied:
OK we have to live with this for now, but there should be a plain data holder for the local variable without the word frame in its name.

cdumez committed:
Adopt even more smart pointers in DOM code
https://bugs.webkit.org/show_bug.cgi?id=263505

Reviewed by Darin Adler.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWindowProxy const):
(WebCore::Document::checkedSelection):
(WebCore::Document::checkedSelection const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentParser.cpp:
(WebCore::DocumentParser::protectedDocument const):
* Source/WebCore/dom/DocumentParser.h:
* Source/WebCore/dom/PendingScript.h:
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::Range):
(WebCore::Range::~Range):
(WebCore::Range::protectedOwnerDocument):
(WebCore::Range::updateAssociatedSelection):
(WebCore::Range::updateAssociatedHighlight):
(WebCore::Range::updateDocument):
(WebCore::childOfCommonRootBeforeOffset):
(WebCore::Range::protectedStartContainer const):
(WebCore::Range::protectedEndContainer const):
(WebCore::Range::processContents):
(WebCore::processContentsBetweenOffsets):
(WebCore::processAncestorsAndTheirSiblings):
(WebCore::Range::insertNode):
(WebCore::Range::toString const):
(WebCore::Range::createContextualFragment):
(WebCore::Range::checkNodeOffsetPair):
(WebCore::Range::cloneRange const):
(WebCore::Range::setStartAfter):
(WebCore::Range::setEndBefore):
(WebCore::Range::setEndAfter):
(WebCore::Range::selectNode):
(WebCore::Range::surroundContents):
(WebCore::Range::setStartBefore):
(WebCore::Range::updateRangeForParentlessNodeMovedToNewDocument):
(WebCore::boundaryTextNodesMerged):
(WebCore::boundaryTextNodesSplit):
(WebCore::Range::expand):
(WebCore::Range::getClientRects const):
(WebCore::Range::getBoundingClientRect const):
(WebCore::setBothEndpoints):
(WebCore::makeSimpleRange):
(WebCore::createLiveRange):
* Source/WebCore/dom/Range.h:
* Source/WebCore/dom/RangeBoundaryPoint.h:
(WebCore::RangeBoundaryPoint::set):
(WebCore::RangeBoundaryPoint::setToAfterNode):
* Source/WebCore/dom/RawDataDocumentParser.h:
* Source/WebCore/dom/RejectedPromiseTracker.cpp:
(WebCore::RejectedPromiseTracker::reportUnhandledRejections):
(WebCore::RejectedPromiseTracker::reportRejectionHandled):
* Source/WebCore/dom/ScopedEventQueue.cpp:
(WebCore::ScopedEventQueue::dispatchEvent const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ref const):
(WebCore::ScriptElement::deref const):
(WebCore::ScriptElement::ref): Deleted.
(WebCore::ScriptElement::deref): Deleted.
* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/ScriptRunner.cpp:
(WebCore::ScriptRunner::~ScriptRunner):
(WebCore::ScriptRunner::queueScriptForExecution):
(WebCore::ScriptRunner::resume):
(WebCore::ScriptRunner::notifyFinished):
(WebCore::ScriptRunner::timerFired):
* Source/WebCore/dom/ScriptRunner.h:
* Source/WebCore/dom/ScriptableDocumentParser.cpp:
(WebCore::ScriptableDocumentParser::scriptsWaitingForStylesheetsExecutionTimerFired):
* Source/WebCore/dom/ScriptedAnimationController.cpp:
(WebCore::ScriptedAnimationController::registerCallback):
(WebCore::ScriptedAnimationController::cancelCallback):
(WebCore::ScriptedAnimationController::serviceRequestAnimationFrameCallbacks):
(WebCore::ScriptedAnimationController::scheduleAnimation):
(WebCore::ScriptedAnimationController::protectedDocument):
* Source/WebCore/dom/ScriptedAnimationController.h:
* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::inheritPolicyContainerFrom):
(WebCore::SecurityContext::checkedContentSecurityPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::insertedIntoAncestor):
(WebCore::ShadowRoot::removedFromAncestor):
(WebCore::ShadowRoot::moveShadowRootToNewParentScope):
(WebCore::ShadowRoot::removeAllEventListeners):
(WebCore::ShadowRoot::findAssignedSlot):
(WebCore::ShadowRoot::assignedNodesForSlot):
(WebCore::assignedShadowRootsIfSlotted):
(): Deleted.
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::makeRangeSelectingNode):
* Source/WebCore/dom/SimulatedClick.cpp:
(WebCore::simulateMouseEvent):
(WebCore::simulateClick):
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::addSlotElementByName):
(WebCore::NamedSlotAssignment::didChangeSlot):
(WebCore::NamedSlotAssignment::assignSlots):
(WebCore::ManualSlotAssignment::addSlotElementByName):
(WebCore::ManualSlotAssignment::slotManualAssignmentDidChange):
(WebCore::ManualSlotAssignment::didRemoveManuallyAssignedNode):
* Source/WebCore/dom/StaticRange.cpp:
(WebCore::StaticRange::create):
* Source/WebCore/dom/StringCallback.cpp:
(WebCore::StringCallback::scheduleCallback):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::setInlineStyleFromString):
(WebCore::StyledElement::invalidateStyleAttribute):
(WebCore::StyledElement::addSubresourceAttributeURLs const):
* Source/WebCore/dom/TagCollection.cpp:
(WebCore::TagCollectionNS::~TagCollectionNS):
(WebCore::TagCollection::~TagCollection):
(WebCore::HTMLTagCollection::~HTMLTagCollection):
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
* Source/WebCore/dom/Text.cpp:
(WebCore::Text::splitText):
(WebCore::Text::replaceWholeText):
(WebCore::Text::virtualCreate):
(WebCore::Text::updateRendererAfterContentChange):
(WebCore::Text::setDataAndUpdate):
* Source/WebCore/dom/TextDecoderStreamDecoder.cpp:
(WebCore::TextDecoderStreamDecoder::decode):
(WebCore::TextDecoderStreamDecoder::flush):
(WebCore::TextDecoderStreamDecoder::protectedTextDecoder):
* Source/WebCore/dom/TextDecoderStreamDecoder.h:
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.h:

Canonical link: https://commits.webkit.org/269644@main
